src/lib.rs

```diff
@@ -124,6 +124,21 @@ pub trait Field:
 
         res
     }
+
+    /// Returns `pairs.into_iter().fold(Self::zero(), |acc, (a_i, b_i)| acc + a_i * b_i)`.
+    ///
+    /// This computes the "dot product" or "inner product" `a ⋅ b` of two equal-length
+    /// sequences of elements `a` and `b`, such that `pairs = a.zip(b)`.
+    ///
+    /// The provided implementation of this trait method uses the direct calculation given
+    /// above. Implementations of `Field` should override this to use more efficient
+    /// methods that take advantage of their internal representation, such as interleaving
+    /// or sharing modular reductions.
+    fn sum_of_products<'a, I: IntoIterator<Item = (&'a Self, &'a Self)> + Clone>(pairs: I) -> Self {
```

Comment:
I originally made this an iterator of `&(a, b)`, and I tested this against the array impl and saw little-to-no performance difference. I then changed it to an iterator of `(&a, &b)` because I thought it would be more flexible, but doing that significantly hurts performance (I presume because we're calling `.clone()` on an owned tuple that contains references, rather than on a reference).

Reply:
Yeah, even using nicer tricks for constructing the input arrays, having this API take `(&a, &b)` causes `bls12_381` pairings to be over 5% slower compared to arrays or `&(a, b)`.

Reply:
Tried two iterators instead:
```
fn sum_of_products<'a>(
        a: impl IntoIterator<Item = &'a Self> + Clone,
        b: impl IntoIterator<Item = &'a Self> + Clone,
    ) -> Self
```
and then `a.clone().into_iter().zip(b.clone().into_iter())`; this is similarly slower than arrays.